### ISSUE TITLE
Type checker accepts tuple returns for functions with "tuple | OtherType" return annotations

### ISSUE THREAD

**Issue by BaconPancakes** [microsoft/pyright#9245]
Is the below behavior expected?

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCqKSYKAsAFCUAmApsFBHAPrACuKAxgBQCUUAWgB8sNggA2tANqoYAGigA6ZQF0oAH0woYALkpQDUELRhsQKURNrcQAQ3TWAzL34BiKAGUAFmDbjqKADk%2BDBeSADOUABGtFD2ULQg4CAA-JQ09Iws7FzMHCQofIIiRAVSplYy2grKiioKsip6FIZGJmYWFZI29mhOLukUdAxMrByczKjAicbUfDpQ7lMztNSWklXySqr6hsam5uvWdg7czryUQA)

```python
from typing import Union

def my_func() -> tuple[int, ...] | int:
    return tuple(range(3)) # Shouldn't this be an error?

def my_func_union() -> Union[tuple[int, ...], int]:
    return tuple(range(3))

def my_func_inferred(): # inferred tuple[int, ...]
    return tuple(range(3))

```

**VS Code extension or command-line**
Pylance language server 2024.10.1 (pyright version 1.1.382, commit e46efc6b)

**Comment by erictraut** [comment]
Pyright's behavior is correct here and conformant with the [Python typing spec](https://typing.readthedocs.io/en/latest/spec/index.html).

The type annotation `tuple[int, ...]` is effectively a union of all tuples that contain only `int` elements of length zero or more. In other words, `tuple[int, ...]` is equivalent to `tuple[()] | tuple[int] | tuple[int, int] | ...`.

For more details, refer to [the tuples chapter](https://typing.readthedocs.io/en/latest/spec/tuples.html) of the typing spec.
